pkg/git/git_test.go

```diff
@@ -17,563 +17,77 @@ limitations under the License.
-	testRepo := newTestRepo(t)
-	defer testRepo.cleanup(t)
-
-	masterRev, err := testRepo.sut.RevParseShort(Master)
-	require.Nil(t, err)
-	require.Equal(t, masterRev, testRepo.firstCommit[:10])
-
-	branchRev, err := testRepo.sut.RevParseShort(testRepo.branchName)
-	require.Nil(t, err)
-	require.Equal(t, branchRev, testRepo.secondBranchCommit[:10])
-
-	tagRev, err := testRepo.sut.RevParseShort(testRepo.firstTagName)
-	require.Nil(t, err)
-	require.Equal(t, tagRev, testRepo.firstCommit[:10])
-}
-
-func TestFailureRevParseShort(t *testing.T) {
-	testRepo := newTestRepo(t)
-	defer testRepo.cleanup(t)
-
-	_, err := testRepo.sut.RevParseShort("wrong")
-	require.NotNil(t, err)
-}
-
-func TestSuccessPush(t *testing.T) {
-	testRepo := newTestRepo(t)
-	defer testRepo.cleanup(t)
-
-	err := testRepo.sut.Push(Master)
-	require.Nil(t, err)
-}
-
-func TestFailurePush(t *testing.T) {
-	testRepo := newTestRepo(t)
-	defer testRepo.cleanup(t)
-
-	err := testRepo.sut.Push("wrong")
-	require.NotNil(t, err)
-}
-
-func TestSuccessRemotify(t *testing.T) {
-	newRemote := Remotify(Master)
-	require.Equal(t, newRemote, DefaultRemote+"/"+Master)
-}
-
-func TestSuccessIsReleaseBranch(t *testing.T) {
-	require.True(t, IsReleaseBranch("release-1.17"))
-}
-
-func TestFailureIsReleaseBranch(t *testing.T) {
-	require.False(t, IsReleaseBranch("wrong-branch"))
-}
-
-func TestSuccessLatestTagForBranch(t *testing.T) {
-	testRepo := newTestRepo(t)
-	defer testRepo.cleanup(t)
-
-	version, err := testRepo.sut.latestTagForBranch(Master)
-	require.Nil(t, err)
-	require.Equal(t, util.AddTagPrefix(version.String()), testRepo.firstTagName)
-}
-
-func TestFailureLatestTagForBranchInvalidBranch(t *testing.T) {
-	testRepo := newTestRepo(t)
-	defer testRepo.cleanup(t)
-
-	version, err := testRepo.sut.latestTagForBranch("wrong-branch")
-	require.NotNil(t, err)
-	require.Nil(t, version)
-}
-
-func TestSuccessLatestPatchToPatch(t *testing.T) {
-	testRepo := newTestRepo(t)
-	defer testRepo.cleanup(t)
-
-	nextMinorTag := "v1.17.1"
-	require.Nil(t, command.NewWithWorkDir(
-		testRepo.sut.Dir(), "git", "tag", nextMinorTag,
-	).RunSuccess())
-
-	result, err := testRepo.sut.LatestPatchToPatch(testRepo.branchName)
-	require.Nil(t, err)
-	require.Equal(t, result.StartSHA(), testRepo.firstCommit)
-	require.Equal(t, result.StartRev(), testRepo.firstTagName)
-	require.Equal(t, result.EndRev(), nextMinorTag)
-}
-
-func TestFailureLatestPatchToPatchWrongBranch(t *testing.T) {
-	testRepo := newTestRepo(t)
-	defer testRepo.cleanup(t)
-
-	result, err := testRepo.sut.LatestPatchToPatch("wrong-branch")
-	require.NotNil(t, err)
-	require.Equal(t, DiscoverResult{}, result)
+type worktreeMock struct {
+	AddCalled      bool
+	CommitCalled   bool
+	CheckoutCalled bool
 }
 
-func TestSuccessDry(t *testing.T) {
-	testRepo := newTestRepo(t)
-	defer testRepo.cleanup(t)
-
-	testRepo.sut.SetDry()
-
-	err := testRepo.sut.Push(Master)
-	require.Nil(t, err)
+func (r *repoMock) CommitObject(plumbing.Hash) (*object.Commit, error) {
```

Comment:
TL;DR: I'd highly recommend `counterfeiter` (or something similar) over implementing the fakes yourself.

Instead of implementing the fakes yourself, you could use `counterfeiter` (& `go generate`).

E.g. [this](https://github.com/kubernetes/release/blob/master/pkg/notes/client.go#L26-L30) allows you to just run `go generate ./...` and it will create / update [the fake](https://github.com/kubernetes/release/blob/master/pkg/notes/notesfakes/fake_client.go) for you. The fake then has nice methods to check on call counts and call args, stub whole method calls, inject fake return values, and those kinda things. When you are generating fakes for exported interfaces/funcs it will also add a compile time assertion to check if the fake actually implements the interface (to avoid drift). However, you can also fake unexported stuff just fine.

Reply:
Okay cool, I'm now using the `counterfeiter` generated mocks. The thing is that we now have to provide two additional methods to overwrite the `inner` and `worktree` to avoid making them public... I also cannot use the same package because I will get a cyclic import if I do so...